elementary/monitor/alerts/group_of_alerts.py

```diff
@@ -0,0 +1,331 @@
+                )
+        self._message_builder._add_blocks_as_attachments(details_blocks)
+
+        return self._message_builder.get_slack_message()
+
+    def _title_block(self):
+        title = f":small_red_triangle: {self._title} ({len(self.alerts)} alerts)"
+        return self._message_builder.create_header_block(title)
+
+    def _number_of_failed_block(self):
+        # small_red_triangle: Falied: 36    |    :Warning: Warning: 3    |    :exclamation: Errors: 1
+        fields = []
+        all_components = list(self._components_to_alerts.items())
+        all_components_but_last = all_components[:-1]
+        for component, alert_list in all_components_but_last:
+            fields.append(
+                f":{component.emoji_in_summary}: {component.name_in_summary}: {len(alert_list)}    |"
+            )
+        component, alert_list = all_components[-1]
+        fields.append(
+            (
+                f":{component.emoji_in_summary}: {component.name_in_summary}: {len(alert_list)}"
+            )
+        )
+
+        return self._message_builder.create_context_block(fields)
+
+    def _get_banner_block(self, env):
+        raise NotImplementedError
+
+    def _get_model_error_blocks(self) -> List:
+        model_error_alert_list = self._components_to_alerts[ModelErrorComponent]
+        if len(model_error_alert_list) == 0:
+            return []
+        result = []
+        for model_error_alert in model_error_alert_list:
+            if model_error_alert.message:
+                result.extend(
+                    [
+                        self._message_builder.create_context_block(
+                            ["*Result message*"]
+                        ),
+                        self._message_builder.create_text_section_block(
+                            f"```{model_error_alert.message.strip()}```"
+                        ),
+                    ]
+                )
+        return result
+
+    def _attention_required_blocks(self):
+        preview_blocks = []
+
+        for component, val in self._components_to_attention_required.items():
+            text = f"_{component.empty_section_content}_" if not val else val
+            preview_blocks.append(
+                self._message_builder.create_text_section_block(
+                    f"*{component.name_in_summary}*: {text}"
+                )
+            )
+
+        preview_blocks.append(self._message_builder.create_empty_section_block())
+
+        return preview_blocks
+
+    def _tabulate_list_of_alerts(self, alert_list):
+        ret = []
+        for alert in alert_list:
+            ret.append(self._get_tabulated_row_from_alert(alert))
+        return "\n".join(ret)
+
+    def _get_tabulated_row_from_alert(self, alert: Alert):
+        raise NotImplementedError
+
+    def _had_channel_clashes(self):
+        return False
+
+    def _get_title(self):
+        return None
+
+
+class GroupOfAlertsByTable(GroupOfAlerts):
```

Comment:
So if you split the the Slack from the GroupOfAlerts, I think that the code will look like:
```
class GroupOfAlerts
     self.slacke_message_builder = SlackGroupOfAlertsMessageBuilder()

      to_slack()


class GroupOfAlertsByTable(GroupOfAlerts):
     self.slacke_message_builder = SlackGroupOfAlertsByTableMessageBuilder()

      to_slack()

class SlackGroupOfAlertsByTableMessageBuilder(SlackGroupOfAlertsMessageBuilder)
```

This is just a sketch, but I think that something like that could make the separation better and will help us support other systems for alerts.
WDYT?

Reply:
We spoke - decided to take all logic that involves `self._message_builder` to live inside `to_slack()` , but to keep the inner functions that return title, summary, ..., attention blocks -they return only strings and the implementation for `to_ms_teams()` will have to take the strings and format them as teams components.
It's done.